README.md
# Overview of HTTP Parser Library

The HTTP Parser Library (HPL) is a high-performance and versatile library written in C, designed for parsing HTTP requests and responses. It seamlessly integrates with `libuv` for efficient I/O operations and utilizes the `llhttp` library for robust HTTP parsing. Tailored for both client and server-side applications, HPL provides an intuitive API for handling HTTP headers, bodies, and various HTTP attributes. Combined with A Memory Library's memory management capabilities, HPL stands as a reliable solution for handling complex HTTP parsing tasks with minimal memory overhead.

## Dependencies

* [libuv](https://github.com/libuv/libuv) for handling I/O operations.
* [A memory library](https://github.com/contactandyc/a-memory-library) for efficient memory management.

## Installation

### Clone the library and change to the directory

```bash
git clone https://github.com/contactandyc/http-parser-library.git
cd http-parser-library
```

### Build and install the library

```bash
mkdir -p build
cd build
cmake ..
make
make install
```

## An Example


`examples/src/json_svc_request.c`:
```c
#include <stdio.h>
#include <string.h>
#include "http-parser-library/hpl.h"
#include "a-memory-library/aml_pool.h"

// Callback function to handle the parsed HTTP data
void on_body_callback(hpl_t *hpl) {
    // Print the URL
    printf("URL: %.*s\n", (int)hpl->url.len, hpl->url.base);

    // Print each header
    for (hpl_header_t *header = hpl->headers; header != NULL; header = header->next) {
        printf("Header: %.*s: %.*s\n", (int)header->key.len, header->key.base,
                                       (int)header->value.len, header->value.base);
    }

    // Print the body
    printf("Body: %.*s\n", (int)hpl->body.len, hpl->body.base);

    // Print other details
    printf("Content Length: %zu\n", hpl->content_length);
    printf("HTTP Version: %u.%u\n", hpl->http_major, hpl->http_minor);
}

int main(int argc, char *argv[]) {
    // Initialize a memory pool
    aml_pool_t *pool = aml_pool_init(1024);

    // Initialize HTTP parser for service request
    hpl_t *http_parser = hpl_service_init(on_body_callback, 16384);

    // Sample JSON POST request data.  The http_request is altered by
    // the hpl_data call so it must not be constant!
    char *http_request = aml_pool_strdup(pool, "POST /submit HTTP/1.1\r\n"
                               "Host: example.com\r\n"
                               "Content-Type: application/json\r\n"
                               "Content-Length: 28\r\n\r\n"
                               "{\"name\":\"John Doe\",\"age\":30}");

    // Write data to the HTTP parser
    if (!hpl_data(http_parser, http_request, strlen(http_request))) {
        fprintf(stderr, "Error parsing HTTP request.\n");
    }

    // Clean up
    hpl_destroy(http_parser);
    aml_pool_destroy(pool);

    return 0;
}
```

After installing http-parser-library
```build
$ cd examples
$ mkdir -p build
$ cd build
$ cmake ..
$ make
$ ./json_svc_request
URL: /submit
Header: Host: example.com
Header: Content-Type: application/json
Header: Content-Length: 28
Body: {"name":"John Doe","age":30}
Content Length: 28
HTTP Version: 1.1
```

## Core Functions

- **hpl_service_init**: Initializes the HTTP parser for parsing service requests.
- **hpl_client_init**: Initializes the HTTP parser for parsing client responses.
- **hpl_chunk**: Sets functions to handle chunk encoding in HTTP data.
- **hpl_clear**: Resets the HTTP parser for reuse.
- **hpl_data**: Writes data to the HTTP parser for parsing.
- **hpl_destroy**: Destroys the HTTP parser and frees resources.

## HTTP Parsing

- **Handling Headers**: Parses and stores HTTP headers.
- **URL Parsing**: Extracts and decodes the URL in service requests.
- **Body Parsing**: Manages HTTP body data for both requests and responses.

## Advanced Features

- **Content-Length Handling**: Manages the `Content-Length` attribute.
- **HTTP Version**: Captures the HTTP version from requests and responses.
- **Chunked Encoding**: Supports chunked transfer encoding.
- **Keep-Alive**: Manages keep-alive connections.
- **Status Code**: Retrieves the status code in client responses.

This README provides a comprehensive guide to using the HTTP Parser Library, including its dependencies, installation process, a basic usage example, and an overview of its core and advanced features.

NEWS.md
HTTP Parser Library 0.0.1 (2023-09-12)
======================================
* Initial release as a package.


src/hpl.c
/*
Copyright 2019 Andy Curtis

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include "http-parser-library/hpl.h"
#include "a-memory-library/aml_alloc.h"

typedef struct {
  hpl_t http;

  on_http_cb on_chunk_encoding;
  on_http_cb on_chunk;
  on_http_cb on_chunk_complete;
  on_http_cb on_url;

  uv_buf_t status;
  uv_buf_t header_key;
  uv_buf_t header_value;
  int last_was_value;
  char *body_ptr;
  size_t body_len;
  bool body_terminated;
  hpl_header_t *tail;

  aml_pool_marker_t marker;

  llhttp_t parser;
  llhttp_settings_t settings;
} acs_llhttp_ext_t;

#define DEBUG_OUTPUT(at, length) ;

#define DEBUG_OUTPUTX(at, length)       \
    printf("%s\n", __FUNCTION__);       \
    printf("at(%p,%lu): ", at, length); \
    fwrite(at, length, 1, stdout);      \
    printf("\n")

/* llhttp methods */
static inline void save_current_header(acs_llhttp_ext_t *lh) {
  hpl_header_t *h = (hpl_header_t *)aml_pool_alloc(
      lh->http.pool, sizeof(hpl_header_t));
  h->key = lh->header_key;
  h->value = lh->header_value;
  h->next = NULL;
  if (!lh->tail)
    lh->http.headers = lh->tail = h;
  else {
    lh->tail->next = h;
    lh->tail = h;
  }
}

static inline void copy_buf(aml_pool_t *pool, uv_buf_t *b, const char *at,
                            size_t length) {
  b->base = (char *)aml_pool_ualloc(pool, length + 1);
  memcpy(b->base, at, length);
  b->base[length] = 0;
  b->len = length;
}

static inline void set_buf(aml_pool_t *pool, uv_buf_t *b, const char *at,
                           size_t length) {
  if (b->len > 0) {
    char *d = (char *)aml_pool_ualloc(pool, b->len + length + 1);
    memcpy(d, b->base, b->len);
    char *p = d + b->len;
    memcpy(p, at, length);
    p[length] = 0;
    b->base = d;
    b->len += length;
  } else {
    b->base = (char *)at;
    b->len = length;
  }
}

static int on_status(llhttp_t *parser, const char *at, size_t length) {
  DEBUG_OUTPUT(at, length);
  acs_llhttp_ext_t *lh = (acs_llhttp_ext_t *)parser->data;
  set_buf(lh->http.pool, &lh->status, at, length);
  return 0;
}

static int on_header_key(llhttp_t *parser, const char *at, size_t length) {
  DEBUG_OUTPUT(at, length);
  acs_llhttp_ext_t *lh = (acs_llhttp_ext_t *)parser->data;
  if (lh->last_was_value && lh->header_key.len > 0) {
    save_current_header(lh);
    lh->header_key.len = 0;
    lh->header_value.len = 0;
  }
  set_buf(lh->http.pool, &lh->header_key, at, length);
  lh->last_was_value = 0;
  lh->header_value.len = 0;
  return 0;
}

static int on_header_value(llhttp_t *parser, const char *at, size_t length) {
  DEBUG_OUTPUT(at, length);
  acs_llhttp_ext_t *lh = (acs_llhttp_ext_t *)parser->data;
  set_buf(lh->http.pool, &lh->header_value, at, length);
  lh->last_was_value = 1;
  return 0;
}

static int on_url(llhttp_t *parser, const char *at, size_t length) {
  DEBUG_OUTPUT(at, length);
  /* for some reason, this doesn't work quite right without copying URL */
  acs_llhttp_ext_t *lh = (acs_llhttp_ext_t *)parser->data;
  if (!lh->http.url.len)
    copy_buf(lh->http.pool, &lh->http.url, at, length);
  else
    set_buf(lh->http.pool, &lh->http.url, at, length);

  return 0;
}

static int on_body(llhttp_t *parser, const char *at, size_t length) {
  DEBUG_OUTPUT(at, length);
  acs_llhttp_ext_t *lh = (acs_llhttp_ext_t *)parser->data;
  uv_buf_t *b = &lh->http.body;
  if (!b->len) {
    if (lh->http.content_length > length) {
      b->base =
          (char *)aml_pool_alloc(lh->http.pool, lh->http.content_length + 1);
      b->base[lh->http.content_length] = 0;
      b->len = lh->http.content_length;
      memcpy(b->base, at, length);
      lh->body_ptr = b->base + length;
      lh->body_len = lh->http.content_length - length;
      lh->body_terminated = true;
    } else {
      b->base = (char *)at;
      b->len = lh->http.content_length;
      lh->body_terminated = false;
      lh->body_len = 0;
    }
  } else {
    char *ptr = lh->body_ptr;
    if (length < lh->body_len) {
      memcpy(ptr, at, length);
      ptr += length;
      lh->body_len -= length;
      lh->body_ptr = ptr;
    } else {
      if (lh->body_len)
        memcpy(ptr, at, lh->body_len);
      lh->body_len = 0;
    }
  }
  return 0;
}

static int on_headers_complete(llhttp_t *parser) {
  acs_llhttp_ext_t *lh = (acs_llhttp_ext_t *)parser->data;
  DEBUG_OUTPUT(lh->http.url.base, lh->http.url.len);

  if (lh->header_key.len > 0 && lh->header_value.len > 0)
    save_current_header(lh);

  hpl_header_t *head = lh->http.headers;
  while (head) {
    head->key.base[head->key.len] = 0;
    head->value.base[head->value.len] = 0;
    // printf("%s: %s\n", head->key.base, head->value.base);
    head = head->next;
  }

  lh->header_key.len = 0;
  lh->header_value.len = 0;
  lh->http.body.len = lh->body_len = 0;
  lh->http.content_length = parser->content_length;
  lh->http.http_major = parser->http_major;
  lh->http.http_minor = parser->http_minor;
  lh->http.method = parser->method;
  lh->http.status_code = parser->status_code;
  lh->http.keep_alive = llhttp_should_keep_alive(parser);

  lh->http.chunked = ((lh->parser.flags & F_CHUNKED) == 1) ? true : false;
  if (lh->http.chunked) {
    lh->on_chunk_encoding((hpl_t *)lh);
    aml_pool_save(lh->http.pool, &lh->marker);
  }
  return 0;
}

static int on_message_begin(llhttp_t *parser) { return 0; }

static int on_chunk_header(llhttp_t *parser) {
  acs_llhttp_ext_t *lh = (acs_llhttp_ext_t *)parser->data;
  DEBUG_OUTPUT(lh->http.url.base, lh->http.url.len);
  lh->http.chunked = true;
  lh->http.content_length = parser->content_length;
  lh->body_len = 0;
  lh->body_ptr = NULL;
  lh->body_terminated = false;
  lh->http.body.len = 0;
  lh->http.body.base = NULL;
  aml_pool_restore(lh->http.pool, &lh->marker);

  if (parser->content_length == 0) {
    lh->on_chunk_complete((hpl_t *)lh);
    return 0;
  }
  return 0;
}

static int on_message_complete(llhttp_t *h) {
  acs_llhttp_ext_t *lh = (acs_llhttp_ext_t *)h->data;
  DEBUG_OUTPUT(lh->http.url.base, lh->http.url.len);
  if (lh->http.body.len) {
    char *ch = NULL;
    char c;
    if (!lh->body_terminated) {
      ch = lh->http.body.base + lh->http.body.len;
      c = *ch;
      *ch = 0;
    }
    if (!lh->http.chunked)
      lh->on_url((hpl_t *)lh);
    else if (lh->http.chunked)
      lh->on_chunk_complete((hpl_t *)lh);
    if (ch)
      *ch = c;
  } else {
    if (!lh->http.chunked)
      lh->on_url((hpl_t *)lh);
    else if (lh->http.chunked)
      lh->on_chunk_complete((hpl_t *)lh);
  }
  return 0;
}

static int on_chunk_complete(llhttp_t *parser) {
  acs_llhttp_ext_t *lh = (acs_llhttp_ext_t *)parser->data;
  DEBUG_OUTPUT(lh->http.body.base, lh->http.body.len);
  if (lh->http.body.len) {
    char *ch = NULL;
    char c;
    if (!lh->body_terminated) {
      ch = lh->http.body.base + lh->http.body.len;
      c = *ch;
      *ch = 0;
    }
    if (lh->http.chunked)
      lh->on_chunk((hpl_t *)lh);
    if (ch)
      *ch = c;
  }
  lh->http.body.len = 0;
  return 0;
}

static void on_http(hpl_t *h) {}

static hpl_t *hpl_init(size_t pool_size, bool client,
                               on_http_cb _on_url) {
  acs_llhttp_ext_t *h = (acs_llhttp_ext_t *)aml_malloc(sizeof(acs_llhttp_ext_t));
  memset(&(h->http), 0, sizeof(h->http));
  h->http.pool = aml_pool_init(pool_size == 0 ? 4096 : pool_size);
  h->on_url = _on_url ? _on_url : on_http;
  h->on_chunk_encoding = on_http;
  h->on_chunk = on_http;
  h->on_chunk_complete = on_http;
  h->status.len = 0;
  h->status.base = NULL;
  h->header_key.len = 0;
  h->header_key.base = NULL;
  h->header_value.len = 0;
  h->header_value.base = NULL;
  h->last_was_value = 0;
  h->body_len = 0;
  h->body_ptr = NULL;
  h->body_terminated = false;
  h->tail = NULL;
  memset(&h->parser, 0, sizeof(h->parser));
  memset(&h->settings, 0, sizeof(h->settings));

  h->settings.on_status = on_status;
  h->settings.on_header_field = on_header_key;
  h->settings.on_header_value = on_header_value;
  h->settings.on_headers_complete = on_headers_complete;
  h->settings.on_body = on_body;
  h->settings.on_message_begin = on_message_begin;
  h->settings.on_message_complete = on_message_complete;
  h->settings.on_chunk_header = on_chunk_header;
  h->settings.on_chunk_complete = on_chunk_complete;
  h->settings.on_url = on_url;

  llhttp_init(&h->parser, client ? HTTP_RESPONSE : HTTP_REQUEST, &h->settings);
  h->parser.data = h;

  return (hpl_t *)h;
}

hpl_t *hpl_service_init(on_http_cb on_url, size_t pool_size) {
  return hpl_init(pool_size, false, on_url);
}

hpl_t *hpl_client_init(on_http_cb on_body, size_t pool_size) {
  return hpl_init(pool_size, true, on_body);
}

void hpl_clear(hpl_t *h) {
  aml_pool_t *pool = h->pool;
  void *data = h->data;
  memset(h, 0, sizeof(*h));
  h->pool = pool;
  h->data = data;
  aml_pool_clear(pool);
  acs_llhttp_ext_t *hp = (acs_llhttp_ext_t *)h;
  hp->status.len = 0;
  hp->status.base = NULL;
  hp->header_key.len = 0;
  hp->header_key.base = NULL;
  hp->header_value.len = 0;
  hp->header_value.base = NULL;
  hp->last_was_value = 0;
  hp->body_len = 0;
  hp->body_ptr = NULL;
  hp->body_terminated = false;
  hp->tail = NULL;
}

bool hpl_data(hpl_t *h, const void *d, size_t len) {
  acs_llhttp_ext_t *hp = (acs_llhttp_ext_t *)h;
  return llhttp_execute(&hp->parser, (const char *)d, len) ? false : true;
}

void hpl_chunk(hpl_t *h, on_http_cb on_chunk,
                   on_http_cb on_chunk_encoding, on_http_cb on_chunk_complete) {
  acs_llhttp_ext_t *hp = (acs_llhttp_ext_t *)h;
  hp->on_chunk_encoding = on_chunk_encoding ? on_chunk_encoding : on_http;
  hp->on_chunk = on_chunk ? on_chunk : on_http;
  hp->on_chunk_complete = on_chunk_complete ? on_chunk_complete : on_http;
}

void hpl_destroy(hpl_t *h) {
  aml_pool_destroy(h->pool);
  aml_free(h);
}

#include "llhttp/llhttp.c"

include/http-parser-library/hpl.h
/*
Copyright 2019 Andy Curtis

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#ifndef _hpl_h
#define _hpl_h

#include <stdbool.h>
#include <uv.h>

#include "a-client-server-library/impl/llhttp.h"

#include "a-memory-library/aml_pool.h"

#ifdef __cplusplus
extern "C" {
#endif

typedef struct hpl_header_s {
  uv_buf_t key;
  uv_buf_t value;
  struct hpl_header_s *next;
} hpl_header_t;

typedef struct {
  aml_pool_t *pool;

  void *data;

  hpl_header_t *headers;
  // only for servers
  uv_buf_t url;
  uv_buf_t body;

  size_t content_length;
  unsigned short http_major;
  unsigned short http_minor;

  // only for servers
  unsigned char method;

  // only for clients
  int status_code;

  int keep_alive;
  bool chunked;
} hpl_t;

typedef void (*on_http_cb)(hpl_t *lh);

/* init the http parser for parsing a service request.  The pool_size
   should be zero for default size or a user specified size. */
hpl_t *hpl_service_init(on_http_cb on_url, size_t pool_size);

/* init the http parser for parsing a client response.  The pool_size
   should be zero for default size or a user specified size. */
hpl_t *hpl_client_init(on_http_cb on_body, size_t pool_size);

/* set functions to handle chunk encoding.  on_chunk is typically the only
   function required.  Specify NULL if no callback is specified. */
void hpl_chunk(hpl_t *h, on_http_cb on_chunk,
                   on_http_cb on_chunk_encoding, on_http_cb on_chunk_complete);

/* make the http parser reusable */
void hpl_clear(hpl_t *h);

/* write data to the http parser */
bool hpl_data(hpl_t *h, const void *d, size_t len);

/* destroy the http parser */
void hpl_destroy(hpl_t *h);

#ifdef __cplusplus
}
#endif

#endif


examples/src/json_svc_request.c
#include <stdio.h>
#include <string.h>
#include "http-parser-library/hpl.h"
#include "a-memory-library/aml_pool.h"

// Callback function to handle the parsed HTTP data
void on_body_callback(hpl_t *hpl) {
    // Print the URL
    printf("URL: %.*s\n", (int)hpl->url.len, hpl->url.base);

    // Print each header
    for (hpl_header_t *header = hpl->headers; header != NULL; header = header->next) {
        printf("Header: %.*s: %.*s\n", (int)header->key.len, header->key.base,
                                       (int)header->value.len, header->value.base);
    }

    // Print the body
    printf("Body: %.*s\n", (int)hpl->body.len, hpl->body.base);

    // Print other details
    printf("Content Length: %zu\n", hpl->content_length);
    printf("HTTP Version: %u.%u\n", hpl->http_major, hpl->http_minor);
}

int main(int argc, char *argv[]) {
    // Initialize a memory pool
    aml_pool_t *pool = aml_pool_init(1024);

    // Initialize HTTP parser for service request
    hpl_t *http_parser = hpl_service_init(on_body_callback, 16384);

    // Sample JSON POST request data.  The http_request is altered by
    // the hpl_data call so it must not be constant!
    char *http_request = aml_pool_strdup(pool, "POST /submit HTTP/1.1\r\n"
                               "Host: example.com\r\n"
                               "Content-Type: application/json\r\n"
                               "Content-Length: 28\r\n\r\n"
                               "{\"name\":\"John Doe\",\"age\":30}");

    // Write data to the HTTP parser
    if (!hpl_data(http_parser, http_request, strlen(http_request))) {
        fprintf(stderr, "Error parsing HTTP request.\n");
    }

    // Clean up
    hpl_destroy(http_parser);
    aml_pool_destroy(pool);

    return 0;
}
